### PR TITLE
Make NewDeployment wait timeout configurable

### PR DESCRIPTION
**Pull request by vadasambar**
details
-----------
- add "deploytimeout" flag to specify timeout in seconds

Fix for #1213

---
This change is [Reviewable](https://reviewable.io/reviews/fission/fission/1260)

### CODE REVIEW

File: pkg/executor/newdeploy/newdeploy.go [hunk 1]
```diff
@@ -428,7 +429,10 @@ func (deploy *NewDeploy) waitForDeploy(depl *v1beta1.Deployment, replicas int32)
 		}
 		time.Sleep(time.Second)
 	}
-	return nil, errors.New("failed to create deployment within timeout window")
+
+	// this error appears in the executor pod logs
+	timeoutError := fmt.Sprintf("failed to create deployment within timeout window of %d seconds", waitTimeOut)
```
Comment (bhavin192) [inline pkg/executor/newdeploy/newdeploy.go]: You can use `fmt.Errorf` here.

Comment (vadasambar) [reply pkg/executor/newdeploy/newdeploy.go]: Makes sense. I have made the change. Thank you :)

**Comment by vadasambar** [comment]
If there are any more changes, please let me know. Thank you :)

**Comment by vadasambar** [comment]
Travis Build has failed due to network timeout. Can someone please trigger a rebuild?

```sh
Timeout waiting for network availability.
```

**Comment by life1347** [comment]
From the test error log, looks like we have to comment out the validation part and shows warning for now to prevent compatibility issue.

```
test_specs/test_spec_merge/test_spec_merge.sh.log
========== start test_specs/test_spec_merge/test_spec_merge.sh.log ==========
~/gopath/src/github.com/fission/fission/test/tests/test_specs/test_spec_merge ~/gopath/src/github.com/fission/fission
Warning: You have provided both - container spec and pod spec and while merging the pod spec will take precedence.
Warning: You have provided both - container spec and pod spec and while merging the pod spec will take precedence.
Warning: Poolsize can only be configured when environment version equals to 3, default poolsize 3 will be used for creating environment pool.
Fatal error: Failed to validate specs: 1 error occurred:
	* ExecutionStrategy.SpecializationTimeout: Invalid value: 0: [SpecializationTimeout must be a value equal to or greater than 120]
Cleaning up...
~/gopath/src/github.com/fission/fission
Fatal error: Spec directory specs doesn't exist. Please check directory path or run "fission spec init" to create it.
```

Here is the diff I made, you may be able to apply it directly. Changes in diff
1. Comment out specializationtimeout validation 
2. Shows warning when running `fission spec apply` without aborting the task.

```
diff --git a/pkg/apis/fission.io/v1/validation.go b/pkg/apis/fission.io/v1/validation.go
index 085a9ca8..4eddaf6d 100644
--- a/pkg/apis/fission.io/v1/validation.go
+++ b/pkg/apis/fission.io/v1/validation.go
@@ -347,9 +347,10 @@ func (es ExecutionStrategy) Validate() error {
 			result = multierror.Append(result, MakeValidationErr(ErrorInvalidValue, "ExecutionStrategy.TargetCPUPercent", es.TargetCPUPercent, "TargetCPUPercent must be a value between 1 - 100"))
 		}
 
-		if es.SpecializationTimeout < 120 {
-			result = multierror.Append(result, MakeValidationErr(ErrorInvalidValue, "ExecutionStrategy.SpecializationTimeout", es.SpecializationTimeout, "SpecializationTimeout must be a value equal to or greater than 120"))
-		}
+		// TODO Add validation warning
+		//if es.SpecializationTimeout < 120 {
+		//	result = multierror.Append(result, MakeValidationErr(ErrorInvalidValue, "ExecutionStrategy.SpecializationTimeout", es.SpecializationTimeout, "SpecializationTimeout must be a value equal to or greater than 120"))
+		//}
 	}
 
 	return result.ErrorOrNil()
diff --git a/pkg/fission-cli/spec.go b/pkg/fission-cli/spec.go
index ec9ef25a..736aa729 100644
--- a/pkg/fission-cli/spec.go
+++ b/pkg/fission-cli/spec.go
@@ -437,6 +437,10 @@ func (fr *FissionResources) validate(c *cli.Context) error {
 		if _, ok := environments[fmt.Sprintf("%s:%s", f.Spec.Environment.Name, f.Spec.Environment.Namespace)]; !ok {
 			log.Warn(fmt.Sprintf("Environment %s is referenced in function %s but not declared in specs", f.Spec.Environment.Name, f.Metadata.Name))
 		}
+		strategy := f.Spec.InvokeStrategy.ExecutionStrategy
+		if strategy.ExecutorType == fv1.ExecutorTypeNewdeploy && strategy.SpecializationTimeout < DEFAULT_SPECIALIZATION_TIMEOUT {
+			log.Warn(fmt.Sprintf("SpecializationTimeout in function spec.InvokeStrategy.ExecutionStrategy should be a value equal to or greater than %v", DEFAULT_SPECIALIZATION_TIMEOUT))
+		}
 	}
 
 	// (ErrorOrNil returns nil if there were no errors appended.)

```

**Comment by vadasambar** [comment]
I could apply the diff without any problems.

**Comment by life1347** [comment]
CI result with kubernetes integration test: https://travis-ci.org/fission/fission/builds/571724467
LGTM, I will merge this PR later today.

File: pkg/fission-cli/function.go [hunk 2]
```diff
@@ -556,6 +572,19 @@ func fnUpdate(c *cli.Context) error {
 		log.Fatal(err)
 	}
 	function.Spec.InvokeStrategy = *strategy
+
+	if c.IsSet("specializationtimeout") {
+		if c.String("executortype") != types.ExecutorTypeNewdeploy {
+			log.Fatal("specializationtimeout must be greater than or equal to 120 seconds")
```
Comment (life1347) [inline pkg/fission-cli/function.go]: Sorry, I suddenly notice that the message here is wrong, please correct it.

File: pkg/fission-cli/main.go [hunk 3]
```diff
@@ -94,6 +94,7 @@ func NewCliApp() *cli.App {
 	minScale := cli.IntFlag{Name: "minscale", Usage: "Minimum number of pods (Uses resource inputs to configure HPA)"}
 	maxScale := cli.IntFlag{Name: "maxscale", Usage: "Maximum number of pods (Uses resource inputs to configure HPA)"}
 	targetcpu := cli.IntFlag{Name: "targetcpu", Usage: "Target average CPU usage percentage across pods for scaling"}
+	specializationTimeoutFlag := cli.IntFlag{Name: "specializationtimeout, st", Usage: "Timeout for newdeploy function creation"}
```
Comment (life1347) [inline pkg/fission-cli/main.go]:
```
Usage: "Timeout for newdeploy to wait for function pod creation"
```

File: pkg/fission-cli/function.go (hunk 2 above)
Comment (vadasambar) [reply pkg/fission-cli/function.go]: I should have noticed that. Thank you for pointing it out.

File: pkg/fission-cli/main.go (hunk 3 above)
Comment (vadasambar) [reply pkg/fission-cli/main.go]: Fixed

**Comment by life1347** [comment]
Merged! @vadasambar Thanks for the PR! And the changes' detail of each commit is great.

**Comment by vadasambar** [comment]
Thank you :)

**Pull request merged**